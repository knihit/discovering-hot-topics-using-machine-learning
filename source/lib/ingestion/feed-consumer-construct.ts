#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/


import { IStream, StreamEncryption } from '@aws-cdk/aws-kinesis';
import { Code, Function, Runtime, StartingPosition } from '@aws-cdk/aws-lambda';
import { Construct, Duration } from '@aws-cdk/core';
import { KinesisStreamsToLambda } from '@aws-solutions-constructs/aws-kinesisstreams-lambda';
export interface FeedConsumerProps {
    readonly environment?: { [key: string]: string }
    readonly timeout: Duration,
    readonly runtime: Runtime,
    readonly code: Code,
    readonly batchSize?: number,
    readonly retryAttempts?: number,
    readonly shardCount?: number,
    readonly retentionPeriod?: Duration
}

export class FeedConsumer extends Construct {

    private readonly feedConsumerStream: IStream;
    private _consumerFunction: Function;

    constructor(scope: Construct, id: string, props: FeedConsumerProps) {
        super(scope, id);

        const ingestionStream = new KinesisStreamsToLambda(this, 'IngestionStream', {
            lambdaFunctionProps: {
                runtime: props.runtime,
                handler: 'index.handler',
                code: props.code,
                timeout: Duration.minutes(5),
                environment: props.environment
            },
            kinesisStreamProps: {
                encryption: StreamEncryption.MANAGED,
                retentionPeriod: props.retentionPeriod,
                shardCount: props.shardCount
            },
            kinesisEventSourceProps: {
                startingPosition: StartingPosition.TRIM_HORIZON,
                batchSize: props.batchSize,
                retryAttempts: props.retryAttempts
            }
        });

        this._consumerFunction = ingestionStream.lambdaFunction;
        this.feedConsumerStream = ingestionStream.kinesisStream;
    }

    public get lambdaFunction(): Function {
        return this._consumerFunction;
    }

    public get kinesisStream(): IStream {
        return this.feedConsumerStream;
    }
}
