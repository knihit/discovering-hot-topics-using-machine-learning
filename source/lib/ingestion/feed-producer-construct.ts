#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { AttributeType } from '@aws-cdk/aws-dynamodb';
import { Rule, Schedule } from '@aws-cdk/aws-events';
import { LambdaFunction } from '@aws-cdk/aws-events-targets';
import { Effect, PolicyStatement, Role } from '@aws-cdk/aws-iam';
import { IStream } from '@aws-cdk/aws-kinesis';
import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
import { Aws, Construct, Duration } from '@aws-cdk/core';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';

export interface FeedProducerProps {
    readonly timeout: Duration,
    readonly stream: IStream,
    readonly runtime: Runtime,
    readonly code: Code,
    readonly solutionName: string,
    readonly supportedLang: string,
    readonly ingestFrequency: string,
    readonly queryParameter: string,
    readonly credentialKeyPath: string
}

export class FeedProducer extends Construct {

    private readonly producerFn: Function;

    constructor(scope: Construct, id: string, props: FeedProducerProps) {
        super(scope, id);

        const lambdaSSMPolicy = new PolicyStatement({
            effect: Effect.ALLOW,
            resources: [ props.credentialKeyPath? `arn:${Aws.PARTITION}:ssm:${Aws.REGION}:${Aws.ACCOUNT_ID}:parameter${props.credentialKeyPath}`:
                `arn:${Aws.PARTITION}:ssm:${Aws.REGION}:${Aws.ACCOUNT_ID}:parameter/${props.solutionName}/${Aws.STACK_NAME}/*` ],
            actions: [ 'ssm:GetParameter', 'ssm:PutParameter' ]
        });

        const dynamoDBTable = new LambdaToDynamoDB(this, 'LambdaDDB', {
            lambdaFunctionProps: {
                runtime: props.runtime,
                handler: 'index.handler',
                code: props.code,
                timeout: props.timeout,
                environment: {
                    SOLUTION_NAME: props.solutionName,
                    STACK_NAME: Aws.STACK_NAME,
                    STREAM_NAME: props.stream.streamName,
                    SUPPORTED_LANG: props.supportedLang,
                    QUERY_PARAM: props.queryParameter,
                    CAP_NUM_RECORD: '25',
                    CREDENTIAL_KEY_PATH: props.credentialKeyPath
                }
            },
            dynamoTableProps: {
                partitionKey: {
                    name: "ACCOUNT_IDENTIFIER", // socail media account identifier is the partition key
                    type: AttributeType.STRING
                },
                sortKey: {
                    name: "CREATED_TIMESTAMP",
                    type: AttributeType.STRING
                },
                timeToLiveAttribute: "EXP_DATE"
            }
        });

        props.stream.grantWrite(dynamoDBTable.lambdaFunction.role as Role)
        dynamoDBTable.lambdaFunction.addToRolePolicy(lambdaSSMPolicy);

        const rule = new Rule(this, 'PollFrequency', {
            schedule: Schedule.expression(`${props.ingestFrequency}`)
        });

        rule.addTarget(new LambdaFunction(dynamoDBTable.lambdaFunction));
        this.producerFn = dynamoDBTable.lambdaFunction;
    }

    public get producerFunction(): Function {
        return this.producerFn;
    }
}
