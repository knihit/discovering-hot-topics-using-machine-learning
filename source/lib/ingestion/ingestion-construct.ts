#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/


import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
import { Construct, Duration } from '@aws-cdk/core';
import { FeedConsumer } from './feed-consumer-construct';
import { FeedProducer } from './feed-producer-construct';

export interface IngestionProps {
    readonly solutionName: string,
    readonly ingestFrequency: string,
    readonly queryParameter: string,
    readonly supportedLang: string //comma separated values of language codes
    readonly credentialKeyPath: string
}

export class Ingestion extends Construct {
    private readonly _producerLambdaFunction: Function;
    private readonly _consumerLambdaFunction: Function;

    constructor(scope: Construct, id: string, props: IngestionProps) {
        super(scope, id);

        const feedConsumerlambda = new FeedConsumer(this, 'FeedConsumer', {
            runtime: Runtime.NODEJS_12_X,
            code: Code.fromAsset(`${__dirname}/../../lambda/ingestion-consumer`),
            timeout: Duration.minutes(5),
            batchSize: 5,
            shardCount: 1
        });

        this._consumerLambdaFunction = feedConsumerlambda.lambdaFunction;

        // start creation of Lambda Kinesis producer that fetches social media feed
        this._producerLambdaFunction =  new FeedProducer(this, 'FeedProducerConstruct', {
            timeout: Duration.minutes(5),
            stream: feedConsumerlambda.kinesisStream,
            runtime: Runtime.NODEJS_12_X,
            code: Code.fromAsset(`${__dirname}/../../lambda/ingestion-producer`),
            solutionName: props.solutionName,
            supportedLang: props.supportedLang,
            ingestFrequency: props.ingestFrequency,
            queryParameter: props.queryParameter,
            credentialKeyPath: props.credentialKeyPath
        }).producerFunction;
        // end creation of Lambda Kinesis producer that fetches social media feed
    }

    public get producerLambdaFunc(): Function {
        return this._producerLambdaFunction;
    }

    public get consumerLambdaFunc(): Function {
        return this._consumerLambdaFunction;
    }
}
