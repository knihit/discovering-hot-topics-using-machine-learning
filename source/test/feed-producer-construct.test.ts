/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { Stream, StreamEncryption } from '@aws-cdk/aws-kinesis';
import { Code, Runtime } from "@aws-cdk/aws-lambda";
import { Duration, Stack } from '@aws-cdk/core';
import * as assert from 'assert';
import { FeedProducer } from '../lib/ingestion/feed-producer-construct';


test('Test Lambda with role and scheduler', () => {
    const stack = new Stack();

    const producerStack = new FeedProducer(stack,'FeedProducerConstruct', {
        timeout: Duration.minutes(5),
        stream: new Stream(stack, 'TestStream', {
            encryption: StreamEncryption.MANAGED
        }),
        runtime: Runtime.NODEJS_12_X,
        code: Code.fromAsset(`${__dirname}/../lambda/ingestion-producer`),
        solutionName: 'discovering-hot-topics-using-machine-learning',
        supportedLang: 'de,en,es,it,pt,fr,ja,ko,hi,ar,zh-cn,zh-tw',
        ingestFrequency: '(0/2 * * * ? *)',
        queryParameter: 'Health',
        credentialKeyPath: '/some/dummy/path/test'
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    assert(producerStack.producerFunction !== null);
});