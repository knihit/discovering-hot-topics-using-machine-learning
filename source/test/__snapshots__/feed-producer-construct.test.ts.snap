// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Lambda with role and scheduler 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersb3e6cc1355048ca26f5af63b97ef696d7a1b6f844b119b69a171b0ea7bb6c0a7ArtifactHash7BD648D2": Object {
      "Description": "Artifact hash for asset \\"b3e6cc1355048ca26f5af63b97ef696d7a1b6f844b119b69a171b0ea7bb6c0a7\\"",
      "Type": "String",
    },
    "AssetParametersb3e6cc1355048ca26f5af63b97ef696d7a1b6f844b119b69a171b0ea7bb6c0a7S3Bucket05C40F74": Object {
      "Description": "S3 bucket for asset \\"b3e6cc1355048ca26f5af63b97ef696d7a1b6f844b119b69a171b0ea7bb6c0a7\\"",
      "Type": "String",
    },
    "AssetParametersb3e6cc1355048ca26f5af63b97ef696d7a1b6f844b119b69a171b0ea7bb6c0a7S3VersionKey69B30ECB": Object {
      "Description": "S3 key for asset version \\"b3e6cc1355048ca26f5af63b97ef696d7a1b6f844b119b69a171b0ea7bb6c0a7\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "FeedProducerConstructLambdaDDBDynamoTableAA674720": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "ACCOUNT_IDENTIFIER",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "CREATED_TIMESTAMP",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "ACCOUNT_IDENTIFIER",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "CREATED_TIMESTAMP",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
        "TimeToLiveSpecification": Object {
          "AttributeName": "EXP_DATE",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "FeedProducerConstructLambdaDDBLambdaFunctionFFB177EE": Object {
      "DependsOn": Array [
        "FeedProducerConstructLambdaDDBLambdaFunctionServiceRoleDefaultPolicy6725EDBF",
        "FeedProducerConstructLambdaDDBLambdaFunctionServiceRole65571D6E",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions has the required permission to write CloudWatch Logs. It uses custom policy instead of arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole with more tighter permissions.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersb3e6cc1355048ca26f5af63b97ef696d7a1b6f844b119b69a171b0ea7bb6c0a7S3Bucket05C40F74",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb3e6cc1355048ca26f5af63b97ef696d7a1b6f844b119b69a171b0ea7bb6c0a7S3VersionKey69B30ECB",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersb3e6cc1355048ca26f5af63b97ef696d7a1b6f844b119b69a171b0ea7bb6c0a7S3VersionKey69B30ECB",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "CAP_NUM_RECORD": "25",
            "CREDENTIAL_KEY_PATH": "/some/dummy/path/test",
            "DDB_TABLE_NAME": Object {
              "Ref": "FeedProducerConstructLambdaDDBDynamoTableAA674720",
            },
            "QUERY_PARAM": "Health",
            "SOLUTION_NAME": "discovering-hot-topics-using-machine-learning",
            "STACK_NAME": Object {
              "Ref": "AWS::StackName",
            },
            "STREAM_NAME": Object {
              "Ref": "TestStreamE6F40222",
            },
            "SUPPORTED_LANG": "de,en,es,it,pt,fr,ja,ko,hi,ar,zh-cn,zh-tw",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "FeedProducerConstructLambdaDDBLambdaFunctionServiceRole65571D6E",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 300,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "FeedProducerConstructLambdaDDBLambdaFunctionServiceRole65571D6E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "FeedProducerConstructLambdaDDBLambdaFunctionServiceRoleDefaultPolicy6725EDBF": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Lambda needs the following minimum required permissions to send trace data to X-Ray.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "FeedProducerConstructLambdaDDBDynamoTableAA674720",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            Object {
              "Action": Array [
                "kinesis:ListShards",
                "kinesis:PutRecord",
                "kinesis:PutRecords",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestStreamE6F40222",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:GetParameter",
                "ssm:PutParameter",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/some/dummy/path/test",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FeedProducerConstructLambdaDDBLambdaFunctionServiceRoleDefaultPolicy6725EDBF",
        "Roles": Array [
          Object {
            "Ref": "FeedProducerConstructLambdaDDBLambdaFunctionServiceRole65571D6E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "FeedProducerConstructPollFrequencyA3F49326": Object {
      "Properties": Object {
        "ScheduleExpression": "(0/2 * * * ? *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "FeedProducerConstructLambdaDDBLambdaFunctionFFB177EE",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "FeedProducerConstructPollFrequencyAllowEventRuleFeedProducerConstructPollFrequencyD3A9EC5A9B4319F0": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "FeedProducerConstructLambdaDDBLambdaFunctionFFB177EE",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "FeedProducerConstructPollFrequencyA3F49326",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestStreamE6F40222": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
  },
}
`;
